app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/PeerGroupHandshakeAsyncTask.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Once two peer devices are connected through wifi direct, this task is executed to perform a
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver. Finally, the onPostExecute() of the sender
+ * initiates the file transfer through {@link SenderDeviceAsyncTask} on the sender and using
+ * {@link ReceiverDeviceAsyncTask} on the receiver.
+ * */
+class PeerGroupHandshakeAsyncTask extends AsyncTask<Void, Void, InetAddress> {
+
+  private static final String TAG = "PeerGrpHndshakeAsyncTsk";
+  private final String HANDSHAKE_MESSAGE = "Request Kiwix File Sharing";
+
+  private DeviceListFragment deviceListFragment;
+  private WifiP2pInfo groupInfo;
+
+  public PeerGroupHandshakeAsyncTask(DeviceListFragment deviceListFragment, WifiP2pInfo groupInfo) {
+    this.deviceListFragment = deviceListFragment;
+    this.groupInfo = groupInfo;
+  }
+
+  @Override
+  protected InetAddress doInBackground(Void... voids) {
+
+    try {
+      if(groupInfo.groupFormed && groupInfo.isGroupOwner) {
+        ServerSocket serverSocket = new ServerSocket(PEER_HANDSHAKE_PORT);
+        serverSocket.setReuseAddress(true);
+        Socket server = serverSocket.accept();
+
+        ObjectInputStream objectInputStream = new ObjectInputStream(server.getInputStream());
+        Object object = objectInputStream.readObject();
+
+        // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+        if (object.getClass().equals(String.class) && object.equals(HANDSHAKE_MESSAGE)) {
+          if(BuildConfig.DEBUG) Log.d(TAG, "Client IP address: "+ server.getInetAddress());
+
+          exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream());
+
+          server.close();
+
+          return server.getInetAddress();
+        }
+      }
+      else if(groupInfo.groupFormed) { //&& !groupInfo.isGroupOwner
+        Socket client = new Socket();
+        client.setReuseAddress(true);
+        client.connect((new InetSocketAddress(groupInfo.groupOwnerAddress.getHostAddress(), PEER_HANDSHAKE_PORT)), 15000);
+
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(client.getOutputStream());
+        objectOutputStream.writeObject(HANDSHAKE_MESSAGE); // Send message for the peer device to verify
+
+        exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream());
+
+        client.close();
+
+        return groupInfo.groupOwnerAddress;
+      }
+
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    return null; // So a null is only returned in case of an exception
```

Comment:
return it in the catch so

Reply:
Keeping it only inside `catch` causes a 'missing return statement' error, should I put it at both places?

Reply:
ahhhh
```
if (object.getClass().equals(String.class) && object.equals(HANDSHAKE_MESSAGE)) {
          if(BuildConfig.DEBUG) Log.d(TAG, "Client IP address: "+ server.getInetAddress());
          exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream());
          server.close();
          return server.getInetAddress();
        }
```
this has no else return so if  the `if` is false execution flow hits the final return here. I made the mistake of trusting a comment. Delete the comment and have one return I guess

Reply:
> I made the mistake of trusting a comment

That's on me, but the null is returned only when: 1) A try-catch exception has occurred, or 2) The device selected for file transfer is not running the kiwix-app while accepting wifi-direct connections, so it won't be able to participate in file transfer

Should I correct it to: `// So a null is only returned in case of an exception or receiver device not running kiwix app`?

Reply:
well you could make it much more explicit in the code because comments are generally terrible
```
if (isKiwixHandshake(object)) {
          if(BuildConfig.DEBUG) Log.d(TAG, "Client IP address: "+ server.getInetAddress());
          exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream());
          server.close();
          return server.getInetAddress();
        }
else{
      return null
}
```
And null here is not the best either as discussed previously.
Ideally were this in kotlin we could represent this as a sealed class, this would return a `Result` and result has 2 subclasses `Success(InetAdress)`, `Failure(Throwable)`, this general practice can be seen applied [here](https://proandroiddev.com/railway-oriented-programming-in-kotlin-f1bceed399e5).
Not that I am requesting a change around null here but trying to drill in the lesson that everytime you type `return null` ask yourself if you really need to do it.

Reply:
Also if statements without braces is a style violation. 
I do it sometimes in kotlin but only when using the if statement as an expression which is not possible in java

Reply:
Is it a bad idea to return a null even in the case where you've written the method receiving the returned value and have handled that possibility?

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/ab9f9368f35c33ddce721678ac2326e2d753ae62#diff-b6bbaac42877486339bfdb7cc97c9aa4R63)